### DISCUSSION TITLE
Structure of CProjects

### DISCUSSION THREAD

**Discussion by petermr** [petermr/petermr#33] [Announcements]
# CProjects

The data in our analyses are held as (recursive) directories which are used for many purposes. The software (e.g. `pygetpapers`, `docanalysis`, `py4ami`, `pyamiimage` and others can recognize parts of this complex structure and process it. It is constantly evolving as we take on more tasks and descriptions and examples may be scattered over several repositories.

The only absolutes are that there is a single `CProject` directory (the corpus) which contains a number of files common to the projects and usually a number of `CTree`s, subdirectories for each document. Where documents are retrieved from EuropePMC they usually have `CTree`s of the form: `PMCddddddd` but documents from elsewhere may have names based on titles. To try to avoid confusion, non-CTrees are held in top-level directories prefixed with underscores. (This is messy - we'd prefer the documents to be in a separate directory- but that's how things evolved).

# Examples

A CTree document retrieved by `pygetpapers` with a variety of subfiles, described by # comments.

```
PMC6821779/
├── eupmc_result.json
├── ftpfiles
│   ├── 41467_2019_12863_Fig1_HTML.jpg
│   ├── 41467_2019_12863_Fig2_HTML.jpg
│   ├── 41467_2019_12863_Fig3_HTML.jpg
│   ├── 41467_2019_12863_Fig4_HTML.jpg
│   ├── 41467_2019_12863_MOESM1_ESM.pdf
│   └── 41467_2019_Article_12863.nxml
├── fulltext.pdf
├── fulltext.xml
├── pdfimages
│   ├── image.2.1.130_308.594_676.png
│   ├── image.2.2.412_459.618_661.png
│   ├── image.4.1.84_199.87_155.png
│   ├── image.5.1.169_292.75_190.png
│   ├── image.5.2.329_453.75_190.png
│   └── image.6.1.304_436.510_521.png
├── supplementaryfiles
│   └── 41467_2019_12863_MOESM1_ESM.pdf
└── svg
    ├── fulltext-page.0.svg
    ├── fulltext-page.1.svg
    ├── fulltext-page.2.svg
    ├── fulltext-page.3.svg
    ├── fulltext-page.4.svg
    ├── fulltext-page.5.svg
    └── fulltext-page.6.svg
```

**Comment by roopavsm** [comment]
## Corpus of images and captions from PMC
We are creating a corpus of images and captions to run 'pyamiimage' over them and curate a dictionary of terpene synthase. 

To build a corpus of images and captions, we search for 'terpene synthase' in [PMC](https://www.ncbi.nlm.nih.gov/pmc/) and explore the images section beside it. We find about 1165 images after the search. This section provides information regarding the paper title, caption title and description along with the images.  As we get pathway diagrams, tables and graphs in the result, we manually pick the pathway diagrams and create a list of them. 
https://github.com/petermr/CEVOpen/blob/master/image_links.csv

Later, we created a CProject of PMC in `roopa-testing` branch of `pyamiimage`. This contains a Ctree of PMC IDs having images and captions alongside. This curation was done manually on Github. 56 images and captions have been uploaded so far. 
https://github.com/petermr/pyamiimage/tree/roopa-testing

Points to work on:
- The caption format should be checked.
- Test images using `pyamiimage`.
- Uploading on Github should be automated.

**Comment by petermr** [comment]
# CProject examples

These are `CProjects` that have been created in the past to show some of the files created. The `CTree` contents are omitted except for the top directory name.

# py/getpapers output

If `pygetpapers` is run from scratch the directory tree may look like this (some similar files are elided ([...]):
```
$ tree pygetpapers_simple/
pygetpapers_simple/
├── PMC8879267
│   ├── eupmc_result.json
│   ├── ftpfiles
│   │   ├── plants-11-00541-g001.jpg
[...]
│   │   ├── plants-11-00541-g007.jpg
│   │   ├── plants-11-00541-s001.zip
│   │   └── plants-11-00541.nxml
│   ├── fulltext.pdf
│   ├── fulltext.xml
│   └── supplementaryfiles
│       ├── plants-11-00541-g001.jpg
[...]
│       ├── plants-11-00541-g007.jpg
│       └── plants-11-00541-s001.zip
├── PMC8896935
│   ├── eupmc_result.json
│   ├── fulltext.pdf
│   ├── fulltext.xml
│   └── supplementaryfiles
│       ├── BMRI2022-3523920.001.jpg
│       ├── BMRI2022-3523920.002.jpg
│       └── BMRI2022-3523920.003.jpg
├── PMC8933013
│   ├── eupmc_result.json
│   ├── ftpfiles
│   │   ├── IENZ_37_2022658.nxml
│   │   ├── IENZ_A_2022658_F0001_C.jpg
│   │   ├── IENZ_A_2022658_F0002_C.jpg
│   │   ├── IENZ_A_2022658_F0003_C.jpg
│   │   ├── IENZ_A_2022658_M0001.jpg
│   │   └── IENZ_A_2022658_SM4099.pdf
│   ├── fulltext.pdf
│   ├── fulltext.xml
│   └── supplementaryfiles
│       ├── IENZ_A_2022658_F0001_C.jpg
│       ├── IENZ_A_2022658_F0002_C.jpg
│       ├── IENZ_A_2022658_F0003_C.jpg
│       ├── IENZ_A_2022658_M0001.jpg
│       └── IENZ_A_2022658_SM4099.pdf
├── PMC9095257
│   ├── eupmc_result.json
│   ├── fulltext.pdf
│   ├── fulltext.xml
│   └── supplementaryfiles
│       ├── ga1.jpg
[...]
│       └── gr9.jpg
├── PMC9269492
│   ├── eupmc_result.json
│   ├── fulltext.pdf
│   └── fulltext.xml
└── eupmc_results.json.   # toplevel (note plural)
```
The `CProject` children are thus:
```
ls -l pygetpapers_simple/
total 72
drwxr-xr-x  7 pm286  staff    224 20 Jul 09:47 PMC8879267
drwxr-xr-x  6 pm286  staff    192 20 Jul 09:49 PMC8896935
drwxr-xr-x  7 pm286  staff    224 20 Jul 09:42 PMC8933013
drwxr-xr-x  6 pm286  staff    192 20 Jul 09:45 PMC9095257
drwxr-xr-x  5 pm286  staff    160 20 Jul 09:44 PMC9269492
-rw-r--r--  1 pm286  staff  36457 20 Jul 09:49 eupmc_results.json
```
We'll refer to `PMC*` as the `CTree`s and `eupmc_results.json` as `CProject` metadata.

Note that `CTree`s can have a wide range of identifiers:
* start with `[A-Za-z]`
* may contain `[A-Za-z0-9_]` (i.e. no punctuation, spaces, etc.)
* must be unique within the `CProject`

Raw (author/publisher) names are often long, punctuated, spaced and must be mapped onto shorter IDs.

**Comment by petermr** [comment]
The `schema` for a `getpapers` CProject is therefore.
```
<CProject_name>.      # MUST
    eupmc_results.json? # MAY (especially if not created from EuropePMC)
    <CTree_name>*  # SHOULD

```
There will be other files created by analysis programs. At present these are indicated by `_[A-Za-z0-9_]+` Most will be directories.

NOTE. This is messy and it could be better to have a structure like:
```
my_cproject/
    CTrees/
      PMC124567/
      another_ctree/
    derived_data/
      summary_data/
      plots/
      indexes/
```
BUT this would mean renaming the system files. Maybe this is a good time to do it.
